ouroboros-network/test/Test/Ouroboros/Network/Testnet.hs

```diff
@@ -205,6 +241,404 @@ tracerDiffusionSimWithTimeName ntnAddr =
+                                          -> Property
+prop_peer_selection_action_trace_coverage defaultBearerInfo diffScript =
+  let sim :: forall s . IOSim s Void
+      sim = diffusionSimulation (toBearerInfo defaultBearerInfo)
+                                diffScript
+                                tracersExtraWithTimeName
+                                tracerDiffusionSimWithTimeName
+
+      events :: [PeerSelectionActionsTrace NtNAddr]
+      events = mapMaybe (\case DiffusionPeerSelectionActionsTrace st -> Just st
+                               _                                     -> Nothing
+                        )
+             . Trace.toList
+             . fmap (\(WithTime _ (WithName _ b)) -> b)
+             . withTimeNameTraceEvents
+                @DiffusionTestTrace
+                @NtNAddr
+             . Trace.fromList (MainReturn (Time 0) () [])
+             . fmap (\(t, tid, tl, te) -> SimEvent t tid tl te)
+             . take 125000
+             . traceEvents
+             $ runSimTrace sim
+
+      peerSelectionActionsTraceMap :: PeerSelectionActionsTrace NtNAddr -> String
+      peerSelectionActionsTraceMap (PeerStatusChanged _)             =
+        "PeerStatusChanged"
+      peerSelectionActionsTraceMap (PeerStatusChangeFailure _ ft) =
+        "PeerStatusChangeFailure " ++ show ft
+      peerSelectionActionsTraceMap (PeerMonitoringError _ se)        =
+        "PeerMonitoringError " ++ show se
+      peerSelectionActionsTraceMap (PeerMonitoringResult _ wspt)     =
+        "PeerMonitoringResult " ++ show wspt
+
+      eventsSeenNames = map peerSelectionActionsTraceMap events
+
+   in tabulate "peer selection actions trace" eventsSeenNames
+      True
+
+-- | This test coverage of TracePeerSelection constructors.
+--
+prop_peer_selection_trace_coverage :: AbsBearerInfo
+                                   -> DiffusionScript
+                                   -> Property
+prop_peer_selection_trace_coverage defaultBearerInfo diffScript =
+  let sim :: forall s . IOSim s Void
+      sim = diffusionSimulation (toBearerInfo defaultBearerInfo)
+                                diffScript
+                                tracersExtraWithTimeName
+                                tracerDiffusionSimWithTimeName
+
+      events :: [TracePeerSelection NtNAddr]
+      events = mapMaybe (\case DiffusionPeerSelectionTrace st -> Just st
+                               _                              -> Nothing
+                        )
+             . Trace.toList
+             . fmap (\(WithTime _ (WithName _ b)) -> b)
+             . withTimeNameTraceEvents
+                @DiffusionTestTrace
+                @NtNAddr
+             . Trace.fromList (MainReturn (Time 0) () [])
+             . fmap (\(t, tid, tl, te) -> SimEvent t tid tl te)
+             . take 125000
+             . traceEvents
+             $ runSimTrace sim
+
+      peerSelectionTraceMap :: TracePeerSelection NtNAddr -> String
```

Comment:
This also possibly could imported from the `testlib`.

Reply:
This one is tricky because `TracePeerSelection` comes from `ouroboros-network` and we do not want cyclic dependencies.